### DISCUSSION TITLE
Would it be possible to effectively make my own aliases from _inside_ a cake script?

### DISCUSSION THREAD

**Discussion by gitfool** [cake-build/cake#4076] [Ideas]
After I updated [gitfool/Cake.Dungeon](https://github.com/gitfool/Cake.Dungeon) for Cake v3, I hit the usual issues with incompatible addins, so I looked at my usage more closely and realized I only needed a _single_ alias from them! I subsequently worked around the issues by removing the addins and inlining my own [extensions](https://github.com/gitfool/Cake.Dungeon/blob/9298d94c2f239a7fceddda9fdf7585bd6f7e3715/scripts/extensions.cake#L3-L7) to `ICakeContext`. This works well enough but has the distinct disadvantage that what was an alias available directly (without qualification) from inside a task must now be explicitly qualified with a `Context` namespace, like [`Context.FileReadText`](https://github.com/gitfool/Cake.Dungeon/blob/9298d94c2f239a7fceddda9fdf7585bd6f7e3715/scripts/tasks.cake#L192).

Hence I am now pondering would it be possible to effectively make my own aliases from _inside_ a cake script, without needing a separately compiled addin, such that they can be used elsewhere in the same loaded cake scripts without qualification. That would be really handy for very simple aliases like the ones I inlined.

Obviously this does not work out of the box, unless I'm missing something, so is this a good idea and what changes would be required?

**Comment by devlead** [comment]
Wondering if it's something like what Cake codegen does you're after, that would look something like

```cake
public static string FileReadText(this ICakeContext context, FilePath file)
    => System.IO.File.ReadAllText(file.MakeAbsolute(context.Environment).FullPath);

public string FileReadText(FilePath file)
   => Context.FileReadText(file);

```

One static extension method and one local utilizing local Context property. Making it avail both in script and a static context.

**Reply by gitfool** [reply]
It was under my nose the whole time.  [Perfect!](https://github.com/gitfool/Cake.Dungeon/commit/dbf1cada2db1c19bcd9d95a96278f1b2be1e3484)

**Reply by devlead** [reply]
Having the extension method too, means it'll work in in a static context too, good if you release as a recipe. 

I.e. if you use a static lamda

```csharp 
Setup(static context => Context.FileReadText("data.txt"));
```

But at same time if your not using it, it could just be YAGNI.

Regardless unsure we've got scoping well documented, if not, could be good addition to website.

**Reply by gitfool** [reply]
Good point re recipe. [Tweaked!](https://github.com/gitfool/Cake.Dungeon/commit/3dcc328ec2cb5678828f4b95d2291b22fee0d6ad) 😉
